ships/0039-build-scheduler-opts.md

```diff
@@ -0,0 +1,214 @@
+   should be easy and explain why the community should desire this enhancement.
+3. **Create a PR.** Assign it to folks with expertise in that domain to help
+   sponsor the process. The PR title should be like "SHIP-NNNN: Neat
+   Enhancement Idea", where "NNNN" is the number associated with this SHIP.
+4. **Merge at each milestone.** Merge when the design is able to transition to a new status
+   (provisional, implementable, implemented, etc.). View anything marked as `provisional` as an idea
+   worth exploring in the future, but not accepted as ready to execute. Anything marked as
+   `implementable` should clearly communicate how an enhancement is coded up and delivered. Aim for
+   single topic PRs to keep discussions focused. If you disagree with what is already in a document,
+   open a new PR with suggested changes.
+
+The `Metadata` section above is intended to support the creation of tooling around the enhancement
+process.
+
+<-->
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+TBD
+
+## Summary
+
+Add API options that influece where `BuildRun` pods are scheduled on Kubernetes. This can be
+acomplished through the following mechanisms:
+
+- [Node Selectors](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector)
+- [Affinity/anti-affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity)
+- [Taints and Tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/)
+
+## Motivation
+
+Today, `BuildRun` pods will run on arbitrary nodes - developers, platform engineers, and admins do
+not have the ability to control where a specific build pod will be scheduled. Teams may have
+several motivations for controlling where a build pod is scheduled:
+
+- Builds can be CPU/memory/storage intensive. Scheduling on larger worker nodes with additional
+  memory or compute can help ensure builds succeed.
+- Clusters may have mutiple worker node architectures and even OS (Windows nodes). Container images
+  are by their nature specific to the OS and CPU architecture, and default to the host operating
+  system and architecture. Builds may need to specify OS and architecture through node selectors.
+- Left unchecked, builds may congregate on a set of nodes, impacting overall cluster utilization
+  and stability.
+
+### Goals
+
+- Allow build pods to run on specific nodes using node selectors.
+- Allow build pods to set node affinity/anti-affinity rules.
+- Allow build pods to tolerate node taints.
+- Allow node selection, pod affinity, and taint toleration to be set at the cluster level.
```

Comment:
> Would a global configuration imply that settings on a build and buildrun are forbidden. Or would we add another configuration option for that?

For idea/inspiration: in OpenShift we have two levels of configuration that were called "defaults" and "overrides." "Defaults" were applied if the user/developer did not specify them, "overrides" wrote over what the user specified. I'm inclined to prefer supporting only "defaults" at the cluster level.

> Or do we delegate forbidding things into policy tools or custom webhooks ?

If the need is to forbid users from changing the default, my gut recommendation is to address this via an admission controller like OPA or Kyverno.

@siamaksade curious what your thought is here.

Reply:
I decided to drop the cluster defaults - this can be considered as a follow-up feature.

Regarding blocking usage of these fields - I added it as a non-goal, recommending this be blocked using an admission controler like OPA Gatekeeper or Kyverno.